Source/render.cpp

```diff
@@ -422,306 +539,834 @@ inline void DoRenderLine(BYTE *dst, BYTE *src, size_t n, BYTE *tbl, DWORD mask)
+					dst += overshoot;
+					drawWidth -= overshoot;
+				}
+				src += v;
+			} else {
+				v = -v;
+				if (v > remainingLeftClip) {
+					const auto overshoot = v - remainingLeftClip;
+					dst += overshoot;
+					drawWidth -= overshoot;
+				}
 			}
+			remainingLeftClip -= v;
+			m = (v == MaxMaskShift) ? 0 : (m << v);
 		}
-		if (arch_draw_type == 2 && tile != RT_RTRIANGLE) {
-			c = block_lvid[level_piece_id];
-			if (c == 2 || c == 3) {
-				if (sgOptions.Graphics.bBlendedTransparancy) // Use a fully transparent mask
-					mask = &RightMask_Transparent[TILE_HEIGHT - 1];
-				else
-					mask = &RightMask[TILE_HEIGHT - 1];
+
+		// Draw the non-clipped segment
+		while (drawWidth > 0) {
+			auto v = static_cast<std::int8_t>(*src++);
+			if (v > 0) {
+				if (v > drawWidth) {
+					RenderLine(dst, src, drawWidth, tbl, m);
+					src += v;
+					dst += drawWidth;
+					drawWidth -= v;
+					break;
+				}
+				RenderLine(dst, src, v, tbl, m);
+				src += v;
+			} else {
+				v = -v;
+				if (v > drawWidth) {
+					dst += drawWidth;
+					drawWidth -= v;
+					break;
+				}
 			}
+			dst += v;
+			drawWidth -= v;
+			m = (v == MaxMaskShift) ? 0 : (m << v);
 		}
-	} else if (arch_draw_type && cel_foliage_active) {
-		if (tile != RT_TRANSPARENT) {
-			return;
+
+		// Skip the rest of src line if clipping on the right
+		assert(drawWidth <= 0);
+		skipRestOfTheLine(clip.right + drawWidth);
+	}
+}
+
+void RenderTransparentSquare(BYTE *dst, int dstPitch, const BYTE *src, const DWORD *mask, const BYTE *tbl, Clip clip) {
+	if (clip.width == Width && clip.height == Height) {
+		RenderTransparentSquareFull(dst, dstPitch, src, mask, tbl);
+	} else {
+		RenderTransparentSquareClipped(dst, dstPitch, src, mask, tbl, clip);
+	}
+}
+
+/** Vertical clip for the lower and upper triangles of a diamond tile (L/RTRIANGLE).*/
+struct DiamondClipY {
+	std::int_fast16_t lowerBottom;
+	std::int_fast16_t lowerTop;
+	std::int_fast16_t upperBottom;
+	std::int_fast16_t upperTop;
+};
+
+template <std::int_fast16_t UpperHeight = TriangleUpperHeight>
+DiamondClipY CalculateDiamondClipY(const Clip &clip)
+{
+	DiamondClipY result;
+	if (clip.bottom > LowerHeight) {
+		result.lowerBottom = LowerHeight;
+		result.upperBottom = clip.bottom - LowerHeight;
+		result.lowerTop = result.upperTop = 0;
+	} else if (clip.top > UpperHeight) {
+		result.upperTop = UpperHeight;
+		result.lowerTop = clip.top - UpperHeight;
+		result.upperBottom = result.lowerBottom = 0;
+	} else {
+		result.upperTop = clip.top;
+		result.lowerBottom = clip.bottom;
+		result.lowerTop = result.upperBottom = 0;
+	}
+	return result;
+}
+
+std::size_t CalculateTriangleSourceSkipLowerBottom(std::int_fast16_t numLines)
+{
+	return XStep * numLines * (numLines + 1) / 2 + 2 * ((numLines + 1) / 2);
+}
+
+std::size_t CalculateTriangleSourceSkipUpperBottom(std::int_fast16_t numLines)
+{
+	return 2 * TriangleUpperHeight * numLines - numLines * (numLines - 1) + 2 * ((numLines + 1) / 2);

```

Comment:
Don't know if this is supposed to do some rounding magic but could also be written as
```suggestion
	return 2 * TriangleUpperHeight * numLines - numLines * numLines + 2 * numLines + 1;
```

Reply:
Same rounding magic as above